src/server/ua_server.c

```diff
@@ -685,25 +685,44 @@ UA_Server_run_iterate(UA_Server *server, UA_Boolean waitInternal) {
         if(hasNext == UA_STATUSCODE_GOOD && multicastNextRepeat < nextRepeated)
             nextRepeated = multicastNextRepeat;
     }
-#endif
-
     UA_UNLOCK(&server->serviceMutex);
+#endif
 
-    now = UA_DateTime_nowMonotonic();
-    timeout = 0;
-    if(nextRepeated > now)
-        timeout = (UA_UInt16)((nextRepeated - now) / UA_DATETIME_MSEC);
     return timeout;
 }
 
 UA_StatusCode
 UA_Server_run_shutdown(UA_Server *server) {
+
     /* Stop the netowrk layer */
     for(size_t i = 0; i < server->config.networkLayersSize; ++i) {
         UA_ServerNetworkLayer *nl = &server->config.networkLayers[i];
         nl->stop(nl, server);
     }
 
+    /* TODO: in future we might want to iterate through the event managers (like net layers) */
+    /* for(size_t i = 0; i < server->config.connectionManagersSize; ++i) {
+        UA_EventSource *eventSource = &server->config.connectionManagers[i].eventSource;
+    */
+    for (size_t i = 0; i < server->config.connectionManagersSize; ++i) {
+        UA_ConnectionManager *cm = server->config.connectionManagers[i];
+        if(cm) {
+
+            if (cm->eventSource.eventLoop != NULL) {
+                cm->eventSource.stop(&cm->eventSource);
+                UA_EventLoop *el = server->config.eventLoop;
+                while(cm->eventSource.state != UA_EVENTSOURCESTATE_STOPPED) {
+                    el->run(el, 0);
```

Comment:
Use a timeout here to prevent spinning.
This can even be UINT32_MAX. Since we don't quite anyway until this is not stopped.